Use a stack to store the TCCL

holly-cummins opened the pull request:
Used for resetting at the end of QuarkusTests and avoid having two code paths both trying to do TCCL-ing.

At first glance, #47657 looked simple. The cause was that we weren't resetting the TCCL back to to the system classloader at the end of quarkus test execution, so plain tests were then being run with the `QuarkusClassLoader`, which did not work well. 

Digging into it, it became clear it was a bit more complex. By the time the `QuarkusTestExtension` methods run, the TCCL is already set to be a runtime classloader, so any resetting is pointless. The pre-set of the TCCL was introduced in https://github.com/quarkusio/quarkus/pull/47379. Even though it was my code it took me a while in the debugger to figure out where it was being set, so I've added some comments. The un-set in the extension listener should work to reset everything, but in nested tests, the 'old' classloader gets overwritten. I've switched to using a stack, which should help with this. 

I've also removed a whole bunch of TCCL-setting in `QuarkusTestExtension`, since we now have the 'super-wrapper' around the whole test which sets and resets the TCCL. Having other TCCL-setting is a no-op and just adds to the code complexity.

- Resolves #47657

holly-cummins committed:
Use a stack for the TCCL, to support nested tests; do not have two code paths both trying to do TCCL-ing

geoand commented:
Thanks a lot for this!

I've relaunched CI in order to see if the failures are related or just flaky tests

holly-cummins commented:
> Thanks a lot for this!
> 
> I've relaunched CI in order to see if the failures are related or just flaky tests

I had a green CI, but then I switched from `Stack` to `Deque`, and added the platform tests. We run the platform ci rather rarely, so it could be something else which has caused the failures (although, let's face it, it's a classloading failure in tests, so this change is the most likely suspect :) ). The spring failure looks like a GitHub CI glitch. 

I'll have a go at reproducing the camel failures locally.

geoand commented:
👌🏽

holly-cummins commented:
Actually, having looked at these failures locally, I'm confident it's unrelated to my changes. I'm not seeing the runtime error, instead I'm seeing a compilation error: 

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.14.0:testCompile (default-testCompile) on project camel-quarkus-integration-test-kubernetes: Compilation failure: Compilation failure: 
[ERROR] /Users/holly/Code/quarkus/camel-quarkus/integration-tests/kubernetes/src/test/java/org/apache/camel/quarkus/component/kubernetes/it/KubernetesConfigMapTest.java:[24,48] cannot find symbol
[ERROR]   symbol:   class KubernetesServer
[ERROR]   location: package io.fabric8.kubernetes.client.server.mock
```

In the CI, `java.lang.NoClassDefFoundError: io/fabric8/kubernetes/client/server/mock/KubernetesServer` happens because `KubernetesServer` has gone away. The commit that moved quarkus to a version of the kubernetes client without that file is https://github.com/quarkusio/quarkus/commit/e114f81cb5d74f589059add21492c9a2fe80294c; camel-quarkus will need updates to adjust to the new file name.

geoand commented on test-framework/junit5/src/main/java/io/quarkus/test/junit/launcher/ExecutionListener.java:
```diff
@@ -44,12 +46,22 @@ public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult
+                if (isQuarkusTest(classLoader)) {
+                    ClassLoader cl = origCl.pop();
+                    if (cl != null) {
+                        System.out.println("HOLLY resetting to " + cl);
```
We probably don't want this :)

holly-cummins replied:
zomg, good spot, thank you!

geoand replied:
🙏🏽

geoand reviewed (approved):
Thanks!

holly-cummins commented:
> @holly-cummins does it make sense to keep the platform tests around if they are just going to be failing until Camel releases? I can imagine this will cause folks some confusion...

I agree it's not ideal. They're only run if people opt-in by setting the `ci/test-platform` label, so most people wouldn't see them. It might be that @gsmet and @jamesnetherton can work out a 'next' branch to run against that would be clean. The Camel team are pro-active about monitoring their version of the platform CI and doing fixes, so the fix is there, just ... not here. :)

geoand commented:
Got it, makes sense.
